From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Johannes Stein <johannesstein@freeze-dev.com>
Date: Fri, 16 Oct 2026 09:50:35 +0100
Subject: [PATCH 54/75] x86 OptPass1ADD/SUB: do not swap past an add/sub that
 reads the register, and merge two SUBs with a negative sum as ADD

---
 compiler/x86/aoptx86.pas | 12 ++++++--
 tests/tbs/tb9052.pp      | 59 ++++++++++++++++++++++++++++++++++++++++
 2 files changed, 68 insertions(+), 3 deletions(-)
 create mode 100644 tests/tbs/tb9052.pp

diff --git a/compiler/x86/aoptx86.pas b/compiler/x86/aoptx86.pas
index 4abf5f3b..fa775fbc 100644
--- a/compiler/x86/aoptx86.pas
+++ b/compiler/x86/aoptx86.pas
@@ -6859,7 +6859,10 @@ unit aoptx86;
                             RemoveInstruction(hp1);
                           end;
                       end
-                    else
+                    else if not RegInOp(ActiveReg, taicpu(hp1).oper[0]^) then
+                      { The second operand must not read the register (e.g.
+                        "addq $1,%rax; addq %rax,%rax" or "addq (%rax),%rax"),
+                        otherwise it would see the value before the addition }
                       begin
                         { Move the constant addition to after the reg/ref addition to improve optimisation }
                         DebugMsg(SPeepholeOptimization + 'Add/sub swap 1a done',p);
@@ -7976,7 +7979,7 @@ unit aoptx86;
                               ) then
                               begin
                                 DebugMsg(SPeepholeOptimization + 'SUB; ADD/SUB -> ADD',p);
-                                taicpu(p).opcode := A_SUB;
+                                taicpu(p).opcode := A_ADD;
                                 taicpu(p).oper[0]^.val := -taicpu(p).oper[0]^.val;
                               end
                             else
@@ -7984,7 +7987,10 @@ unit aoptx86;
                             RemoveInstruction(hp1);
                           end;
                       end
-                    else
+                    else if not RegInOp(ActiveReg, taicpu(hp1).oper[0]^) then
+                      { The second operand must not read the register (e.g.
+                        "subq $3,%rax; subq %rax,%rax"), otherwise it would
+                        see the value before the subtraction }
                       begin
                         { Move the constant subtraction to after the reg/ref addition to improve optimisation }
                         DebugMsg(SPeepholeOptimization + 'Add/sub swap 1b done',p);
diff --git a/tests/tbs/tb9052.pp b/tests/tbs/tb9052.pp
new file mode 100644
index 00000000..804fdcc8
--- /dev/null
+++ b/tests/tbs/tb9052.pp
@@ -0,0 +1,59 @@
+{ %OPT=-O2 }
+{ x86 OptPass1ADD/OptPass1SUB:
+  - "Add/sub swap": "addq $1,%rax; addq %rax,%rax" was reordered to
+    "addq %rax,%rax; addq $1,%rax" although the second instruction reads the
+    register (Inc(x); Inc(x, x) gave 2x+1 instead of 2x+2)
+  - merging "subq $-3,%rax; subq $-4,%rax" into one instruction kept SUB when
+    negating the combined constant (Dec(x, -3); Dec(x, -4) gave x-7) }
+program tb9052;
+
+{$mode objfpc}
+
+function IncInc(x: Int64): Int64; noinline;
+begin
+  Inc(x);
+  Inc(x, x);
+  Result := x;
+end;
+
+function DecDec(x: Int64): Int64; noinline;
+begin
+  Dec(x, 3);
+  Dec(x, x);
+  Result := x;
+end;
+
+function DecNeg(x: Int64): Int64; noinline;
+begin
+  Dec(x, -3);
+  Dec(x, -4);
+  Result := x;
+end;
+
+function IncInc32(x: LongInt): LongInt; noinline;
+begin
+  Inc(x);
+  Inc(x, x);
+  Result := x;
+end;
+
+function DecDec32(x: LongInt): LongInt; noinline;
+begin
+  Dec(x, 3);
+  Dec(x, x);
+  Result := x;
+end;
+
+begin
+  if IncInc(-7)<>-12 then
+    halt(1);
+  if DecDec(10)<>0 then
+    halt(2);
+  if DecNeg(10)<>17 then
+    halt(3);
+  if IncInc32(-7)<>-12 then
+    halt(4);
+  if DecDec32(10)<>0 then
+    halt(5);
+  writeln('ok');
+end.
